### PR TITLE
#3846 - fix for ControlMobVerb not working for ghosts

### LINKED ISSUE #3846: Control Mob verb is broken
## Description

Can't right click control mobs anymore. Nothing happens.

### PR DESCRIPTION
**Pull request by GalacticChimp**
## About the PR 
Fixes #3846 
Because of a previous fix for 'ghosts getting pulled back' #3664 ghosts would be always forbidden from ControlMob'ing bodies. 

A remaining problem is that if you ghost, respawn and ghost again, there will be 2 ghost entities floating around (this problem occurs even without my code changes)

### CODE REVIEW

File: Content.Server/Mobs/Mind.cs [hunk 1]
```diff
@@ -272,8 +275,15 @@ public void TransferTo(IEntity? entity)
             OwnedComponent = component;
             OwnedComponent?.InternalAssignMind(this);
 
-            if (VisitingEntity?.HasComponent<GhostComponent>() == false)
+            GhostComponent? ghostComponent = null;
+            VisitingEntity?.TryGetComponent(out ghostComponent);
+
+            if (ghostComponent == null || // visiting entity is not a Ghost
+                (ghostComponent != null && !ghostComponent.CanReturnToBody) || // it is a ghost, but cannot return to body anyway, so it's okay
+                ghostCheckOverride) // to force mind transfer, for example from ControlMobVerb
```
Comment (GalacticChimp) [inline Content.Server/Mobs/Mind.cs:283]: my best idea was to add an override for the ghost check + add a check if the ghost can actually return to it's body. With this 2 pieces of information it seems to work. Unless I'm missing something.

**Comment by ShadowCommander** [comment]
I haven't had this problem before. What are you doing to reproduce?

**Comment by ShadowCommander** [comment]
I thought I fixed them all in the last ghost PR.

**Comment by mirrorcult** [comment]
this one definitely still happens, reproduced it like 15 minutes ago

**Comment by ShadowCommander** [comment]
I figured it out. Need to ghost from a gibbed mob.

File: Content.Server/Mobs/Mind.cs [hunk 2]
```diff
@@ -272,8 +275,15 @@ public void TransferTo(IEntity? entity)
             OwnedComponent = component;
             OwnedComponent?.InternalAssignMind(this);
 
-            if (VisitingEntity?.HasComponent<GhostComponent>() == false)
+            GhostComponent? ghostComponent = null;
+            VisitingEntity?.TryGetComponent(out ghostComponent);
+
+            if (ghostComponent == null || // visiting entity is not a Ghost
```
Comment (ShadowCommander) [inline Content.Server/Mobs/Mind.cs:281]: You can probably replace some of these checks with the TryGetComponent<GhostComponent>()

File: Content.Server/Mobs/Mind.cs (hunk 1 above)
Comment (ShadowCommander) [inline Content.Server/Mobs/Mind.cs:283]: Put the ghostCheckOverride first.

File: Content.Server/Mobs/Mind.cs [hunk 3]
```diff
@@ -272,8 +275,15 @@ public void TransferTo(IEntity? entity)
...
+                (ghostComponent != null && !ghostComponent.CanReturnToBody) || // it is a ghost, but cannot return to body anyway, so it's okay
+                ghostCheckOverride) // to force mind transfer, for example from ControlMobVerb
+            {
                 VisitingEntity = null;
```
Comment (ShadowCommander) [inline Content.Server/Mobs/Mind.cs:285]: Replacing this with Unvisit() might fix that other issue.

**Review by ShadowCommander — CHANGES_REQUESTED** [review]
There are also a few other places where

**Comment by GalacticChimp** [comment]
> There are also a few other places where
where?

File: Content.Server/Mobs/Mind.cs (hunk 3 above)
Comment (GalacticChimp) [reply Content.Server/Mobs/Mind.cs:284]: it made the screen go black, and then server complained about a missing ActorComponent

File: Content.Server/Mobs/Mind.cs (hunk 1 above)
Comment (GalacticChimp) [reply Content.Server/Mobs/Mind.cs]: done

File: Content.Server/Mobs/Mind.cs [hunk 4]
```diff
@@ -272,8 +275,14 @@ public void TransferTo(IEntity? entity)
             OwnedComponent = component;
             OwnedComponent?.InternalAssignMind(this);
 
-            if (VisitingEntity?.HasComponent<GhostComponent>() == false)
+            GhostComponent? ghostComponent = null;
+
+            if (ghostCheckOverride || // to force mind transfer, for example from ControlMobVerb
+                (VisitingEntity?.TryGetComponent(out ghostComponent) ?? false) || // visiting entity is not a Ghost
+                (ghostComponent != null && !ghostComponent.CanReturnToBody))  // it is a ghost, but cannot return to body anyway, so it's okay
```
Comment (ShadowCommander) [inline Content.Server/Mobs/Mind.cs:282]: This looks a bit off. Try this:
```suggestion
            if (IsVisitingEntity
                && (ghostCheckOverride // to force mind transfer, for example from ControlMobVerb
                || !VisitingEntity.TryGetComponent(out GhostComponent? ghostComponent) // visiting entity is not a Ghost
                || !ghostComponent.CanReturnToBody))  // it is a ghost, but cannot return to body anyway, so it's okay
```
Might need  a not null marker. `!VisitingEntity!.TryGetComponent(out ghostComponent)`

File: Content.Server/Mobs/Mind.cs (hunk 3 above)
Comment (ShadowCommander) [reply Content.Server/Mobs/Mind.cs:284]: Something for a different PR then.

File: Content.Server/Mind/Mind.cs [hunk 5]
```diff
@@ -272,8 +275,15 @@ public void TransferTo(IEntity? entity)
             OwnedComponent = component;
             OwnedComponent?.InternalAssignMind(this);
 
-            if (VisitingEntity?.HasComponent<GhostComponent>() == false)
+#pragma warning disable CS8602 // Dereference of a possibly null reference.
```
Comment (ShadowCommander) [inline Content.Server/Mind/Mind.cs:278]: Remove the disable warning and instead put an exclamation mark next to VisitingEntity like I suggested in the previous review.

**Review by ShadowCommander — CHANGES_REQUESTED** [review]

File: Content.Server/Mind/Mind.cs (hunk 5 above)
Comment (ShadowCommander) [reply Content.Server/Mind/Mind.cs:278]: `!VisitingEntity!.TryGetComponent(out ghostComponent)`

Comment (GalacticChimp) [reply Content.Server/Mind/Mind.cs]: ah yes, sorry I didn't notice the second '!', now everything should be fine

**Review by ShadowCommander — APPROVED** [review]

**Pull request merged**